beamer/util.py

```diff
@@ -136,6 +137,13 @@ def make_web3(
 
 _Token = tuple[ChainId, ChecksumAddress]
 
+
+@dataclass(frozen=True)
+class _TokenData:
+    equivalent_classes: frozenset[_Token]
+    allowance: Optional[int]
```

Comment:
Love that you factored this out!

Nitpick: I would rename `equivalent_classes` to `equivalence_class` because it really refers to the equivalence class of this token (note the singular; https://en.wikipedia.org/wiki/Equivalence_class); it is the equivalence class (just a fancy name for a set) which this token is a member of. 

Also note you can freely use union type syntax instead of `Optional` because we are on a fresh enough Python version: `allowance: None | int`.